### ISSUE TITLE
Avoid xonsh exiting on error

### ISSUE THREAD

**Issue by anki-code** [xonsh/xonsh#3363]
Hello! Thank you for the great shell!

When I try to run some invalid code like that for just experiment:
```
ls | @(lambda a,p: [s for s in p.readline()])
```
The error kicks me out out of xonsh without saving command history:
```
TypeError: '>' not supported between instances of 'str' and 'int'
Traceback (most recent call last):
  File "/root/miniconda3/lib/python3.7/site-packages/xonsh/__amalgam__.py", line 24016, in main
    return main_xonsh(args)
  File "/root/miniconda3/lib/python3.7/site-packages/xonsh/__amalgam__.py", line 24045, in main_xonsh
    shell.shell.cmdloop()
  File "/root/miniconda3/lib/python3.7/site-packages/xonsh/ptk2/shell.py", line 196, in cmdloop
    line = self.singleline(auto_suggest=auto_suggest)
  File "/root/miniconda3/lib/python3.7/site-packages/xonsh/ptk2/shell.py", line 165, in singleline
    line = self.prompter.prompt(**prompt_args)
  File "/root/miniconda3/lib/python3.7/site-packages/prompt_toolkit/shortcuts/prompt.py", line 797, in prompt
    return run_sync()
  File "/root/miniconda3/lib/python3.7/site-packages/prompt_toolkit/shortcuts/prompt.py", line 786, in run_sync
    return self.app.run(inputhook=self.inputhook, pre_run=pre_run2)
  File "/root/miniconda3/lib/python3.7/site-packages/prompt_toolkit/application/application.py", line 736, in run
    return run()
  File "/root/miniconda3/lib/python3.7/site-packages/prompt_toolkit/application/application.py", line 710, in run
    return f.result()
  File "/root/miniconda3/lib/python3.7/site-packages/prompt_toolkit/eventloop/future.py", line 151, in result
    raise self._exception
  File "/root/miniconda3/lib/python3.7/site-packages/prompt_toolkit/eventloop/coroutine.py", line 92, in step_next
    new_f = coroutine.throw(exc)
  File "/root/miniconda3/lib/python3.7/site-packages/prompt_toolkit/application/application.py", line 685, in _run_async2
    result = yield f
  File "/root/miniconda3/lib/python3.7/site-packages/prompt_toolkit/eventloop/coroutine.py", line 88, in step_next
    new_f = coroutine.send(None)
  File "/root/miniconda3/lib/python3.7/site-packages/prompt_toolkit/application/application.py", line 624, in _run_async
    with self.input.raw_mode():
  File "/root/miniconda3/lib/python3.7/site-packages/prompt_toolkit/input/vt100.py", line 231, in __enter__
    termios.tcsetattr(self.fileno, termios.TCSANOW, newattr)
termios.error: (4, 'Interrupted system call')
Xonsh encountered an issue during launch
Failback to /bin/sh

[3]+  Stopped                 xonsh
```
It's unstable behavior for shell. Is there a way to stay at xonsh despite any errors?

**Comment by cafehaine** [comment]
The problem is linked to the "rtns" entry of the history.

It seems that the return value in the history is not limited to an integer, and your prompt bar tries to compare it to an int to determine if there was an error.

I'm not sure what would be the best way to solve this, but the prompt shouldn't compare the return value to 0, and instead check if the return value is different from 0 (as technically, the value could be negative, and there would still be an error).

**Comment by scopatz** [comment]
This works for me:

```
$ ls | @(lambda a,p: [s for s in p.readline()])          
8 
```

What is the output of the `xonfig` command for you?

**Comment by cafehaine** [comment]
It works for me too
```
 ?  18:28 cafehaine@ravenarch ~ master|+1                                      
$ ls | @(lambda a,p: [s for s in p.readline()])                                
📁a
 r  18:28 cafehaine@ravenarch ~ master|+1                                      
$ xonfig                                                                       
+------------------+-----------------+
| xonsh            | 0.9.18          |
| Python           | 3.8.5           |
| PLY              | 3.11            |
| have readline    | True            |
| prompt toolkit   | 3.0.5           |
| shell type       | prompt_toolkit  |
| pygments         | 2.6.1           |
| on posix         | True            |
| on linux         | True            |
| distro           | arch            |
| on darwin        | False           |
| on windows       | False           |
| on cygwin        | False           |
| on msys2         | False           |
| is superuser     | False           |
| default encoding | utf-8           |
| xonsh encoding   | utf-8           |
| encoding errors  | surrogateescape |
+------------------+-----------------+
```
However if I look in the history session (json backend), I find the following line:
```
{"inp": "ls | @(lambda a,p: [s for s in p.readline()])\n", "rtn": "r", "ts": [1596299314.4918275, 1596299314.9915848]
}
```
As you can see, there is a non-integer value in the "rtn" field, which is what I assume caused the crash in the prompt of @anki-code (since by looking at their traceback, there is a comparison between string and integer in the prompt).

**Comment by scopatz** [comment]
Ahh OK! I think that this might be getting fixed by #3402

**Comment by anki-code** [comment]
> When I try to run some invalid code like that for just experiment:
> ls | @(lambda a,p: [s for s in p.readline()])
> The error kicks me out 

Can't repeat in 0.9.21. Closed.
